1-CONTRIBUTION-GUIDE/best-practices.md

```diff
@@ -0,0 +1,252 @@
+
+ ```
+ "parameters": {
+   "storageAccountType": {
+     "type": "string",
+     "metadata": {
+     "description": "The type of the new storage account created to store the VMs disks"
+     }
+   }
+ }
+ ```
+ 
+4. Do not use a parameter to specify the **location**. Use the location property of the resourceGroup instead. By using the **resourceGroup().location** expression for all your resources, the resources in the template will automatically be deployed in the same location as the resource group.
+
+ ```
+ "resources": [
+   {
+     "name": "[variables('storageAccountName')]",
+     "type": "Microsoft.Storage/storageAccounts",
+     "apiVersion": "2015-06-15",
+     "location": "[resourceGroup().location]",
+     "comments": "This storage account is used to store the VM disks",
+     "properties": {
+       "accountType": "Standard_GRS"
+     }
+   }
+ ]
+ ```
+
+5. Name **variables** using this scheme **templateScenarioResourceName** (e.g. simpleLinuxVMVNET, userRoutesNSG, elasticsearchPublicIP etc.) that describe the scenario rather. This ensures when a user browses all the resources in the Portal there aren't a bunch of resources with the same name (e.g. myVNET, myPublicIP, myNSG)
+6. For many resources with a resource group, a name is not often relevant and using something like "storageAccount" may be acceptable.  You can also use variables for the name of a resource. Use **displayName** tags for a "friendly" name in the JSON outline view.  This should ideally match the name property value or property name.
+
+ ```
+ "resources": [
+   {
+     "name": "[variables('storageAccountName')]",
+     "type": "Microsoft.Storage/storageAccounts",
+     "apiVersion": "2015-06-15",
+     "location": "[resourceGroup().location]",
+     "comments": "This storage account is used to store the VM disks",
+     "tags": { "displayName": "storageAccount" },
+     "properties": {
+       "accountType": "Standard_GRS"
+     }
+   }
+ ]
+ ```
+	
+7. Specifying a lower-case **comments** property for each resource in the template, helps other contributors to understand the purpose of the resource.
+
+ ```	
+ "resources": [
+   {
+     "name": "[variables('storageAccountName')]",
+     "type": "Microsoft.Storage/storageAccounts",
+     "apiVersion": "2015-06-15",
+     "location": "[resourceGroup().location]",
+     "comments": "This storage account is used to store the VM disks",
+     "properties": {
+       "accountType": "Standard_GRS"
+     }
+   }
+ ]
+ ```
+
+9. Do not create a parameter for a **storage account name**. Storage account names need to be lower case and can't contain hyphens (-) in addition to other domain name restrictions. A storage account has a limit of 24 characters. They also need to be globally unique. To prevent any validation issue configure a variables (using the expression **uniqueString** and a static value **storage**). Storage accounts with a common prefix (uniquestring) will not get clustered on the same racks.
```

Comment:
@singhkay What is the final guideline?

Option1.
"storageAccountName": "[concat(uniquestring(resourceGroup().id), '<scenario name or a common suffix>')]"

or

Option2.
"storageAccountName": "[concat(uniquestring(resourceGroup().id),'storage')]"

Option2 doesn't require any user input (not error-prone), Option1 one requires functions (toLower and replace) to fix potential issues with the user input. My suggestion would be Option2 as the guideline. If Option1 is required for a template it is accepted too.

Agree?